language/compiler/bytecode_source_map/src/source_map.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use codespan::Span;
+use ir_to_bytecode_syntax::ast::{Loc, QualifiedModuleIdent, Var_};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::ops::Bound;
+use types::account_address::AccountAddress;
+use types::identifier::Identifier;
+use vm::file_format::{CodeOffset, FunctionDefinitionIndex, TableIndex};
+
+//***************************************************************************
+// Source location mapping
+//***************************************************************************
+
+pub type SourceMap = Vec<ModuleSourceMap>;
```

Comment:
The only place I see this type used is in `pub fn source_map_from_file(filename: &str) -> SourceMap`.

Would we want to store more than one `ModuleSourceMap` in a single file? It seems like one map per file (+ a predictable naming convention for the file) might make it easier to use and update source maps produced by by multiple runs of the compiler. For example: imagine you run `movec M1.mvir M2.mvir M3.mvir` and get one file with three maps, then change `M1.mvir` and re-run `movec M1.mvir` to get one file with one map. It wouldn't be clear whether to use the `M1` source map from the first or second run.

By contrast, if you get one source map file per module, the old `M1` source map would get clobbered by the second invocation and there would be no confusion.

Reply:
You do only get one `ModuleSourceMap` per module. However, `CompiledPrograms` can lead to multiple modules being compiled which unless I'm misunderstanding something means that we can have a many-module to single-file mapping. But I may be misunderstanding how programs are represented (and I couldn't find an example readily). 

I agree, if we can have a 1-1 mapping of file to module that would be the best, but I'm not certain that that's possible.  I'll take a look at this tomorrow though and see if we can have this 1-1 setting, because I think this will definitely be the best.

Reply:
Your understanding is totally correct given the state of the code. But as @runtian-zhou mentioned above, we are deprecating `CompiledProgram` and moving to a bytecode file format where there is exactly one module *or* (not and) script per file. So I think we can have that 1-1 mapping!